clippy_lints/src/methods/single_or_empty_collections_iter.rs

```diff
@@ -0,0 +1,205 @@
+use rustc_span::{sym, Span};
+
+use super::{ITER_ON_EMPTY_COLLECTIONS, ITER_ON_SINGLE_ITEMS};
+
+#[derive(Clone, Copy, Debug)]
+enum Variant<'tcx> {
+    SomeToOnce,
+    NoneToEmpty,
+    OneLenToOnce(&'tcx Expr<'tcx>),
+    ZeroLenToEmpty,
+}
+
+impl<'tcx> Variant<'tcx> {
+    fn as_lint(self) -> &'static Lint {
+        match self {
+            Self::SomeToOnce | Self::OneLenToOnce(_) => ITER_ON_SINGLE_ITEMS,
+            Self::NoneToEmpty | Self::ZeroLenToEmpty => ITER_ON_EMPTY_COLLECTIONS,
+        }
+    }
+
+    fn as_str(self) -> &'static str {
+        match self {
+            Self::SomeToOnce => "Some",
+            Self::NoneToEmpty => "None",
+            Self::OneLenToOnce(_) => "[T; 1]",
+            Self::ZeroLenToEmpty => "[T; 0]",
+        }
+    }
+
+    fn desc(self) -> &'static str {
+        match self {
+            Self::SomeToOnce | Self::OneLenToOnce(_) => "iterator with only one element",
+            Self::NoneToEmpty | Self::ZeroLenToEmpty => "empty iterator",
+        }
+    }
+}
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, msrv: &Msrv) {
+    if !msrv.meets(ITER_ONCE_AND_EMPTY) {
+        return;
+    }
+
+    let (variant, is_vec, sugg_span) = match expr.kind {
+        // `[T; 1]`
+        ExprKind::Array([one]) => (Variant::OneLenToOnce(one), false, expr.span),
+        // `[T; 0]`
+        ExprKind::Array([]) => (Variant::ZeroLenToEmpty, false, expr.span),
+        // `Some`
+        ExprKind::Call(path, _) if let Some(def_id) = path_res(cx, path).opt_def_id()
+            && is_lang_item_or_ctor(cx, def_id, LangItem::OptionSome) =>
+        {
+            (Variant::SomeToOnce, false, expr.span)
+        },
+        // `None`
+        ExprKind::Path(qpath) if let Some(def_id) = cx.qpath_res(&qpath, expr.hir_id).opt_def_id()
+            && is_lang_item_or_ctor(cx, def_id, LangItem::OptionNone) =>
+        {
+            (Variant::NoneToEmpty, false, expr.span)
+        }
+        // `vec![]`
+        _ if let Some(mac_call) = root_macro_call_first_node(cx, expr)
+            && let Some(VecArgs::Vec(elems)) = VecArgs::hir(cx, expr) =>
+        {
+            if let [one] = elems {
+                (Variant::OneLenToOnce(one), true, mac_call.span.source_callsite())
+            } else if elems.is_empty() {
+                (Variant::ZeroLenToEmpty, true, mac_call.span.source_callsite())
+            } else {
+                return;
+            }
+        },
+        _ => return,
+    };
+
+    // `vec![]` must be external
+    if !is_vec && in_external_macro(cx.sess(), expr.span) {
+        return;
+    }
+    if let Some(parent) = get_parent_expr(cx, expr)
+        && let ExprKind::MethodCall(path, recv, args, _) = parent
+            .peel_blocks()
+            .peel_borrows()
+            .kind
+    {
+        if recv.hir_id == expr.hir_id
+            && let method_name = path.ident.as_str()
+            && matches!(method_name, "iter" | "iter_mut" | "into_iter")
```

Comment:
`Ident::as_str` is a slow operation (aswell as checking strings), you could check `path.ident` is `sym::iter` or `sym::into_iter`. Maybe I'll make a PR adding `iter_mut` as a symbol in a few days. If this PR is still open by then, it will also be possible to check for them fast.

Reply:
While you're at it, please also add `as_mut_ptr` ^^

Reply:
https://github.com/Centri3/rust-clippy/blob/ebcbc2d73616e9f35990574bfba6aedd02509927/clippy_lints/src/ptr_to_temporary.rs#L347